### PR TITLE
[InstCombine] Fold pointer adding in integer to arithmetic add

### PR DESCRIPTION
**Pull request by YanWQ-monad**
Fold
``` llvm
define i32 @src(i32 %x, i32 %y) {
  %base = inttoptr i32 %x to ptr
  %ptr = getelementptr inbounds i8, ptr %base, i32 %y
  %r = ptrtoint ptr %ptr to i32
  ret i32 %r
}
```
where both `%base` and `%ptr` have only one use, to
``` llvm
define i32 @tgt(i32 %x, i32 %y) {
  %r = add i32 %x, %y
  ret i32 %r
}
```

The `add` can be `nuw` if the GEP is `inbounds` and the offset is non-negative. The relevant Alive2 proof is https://alive2.llvm.org/ce/z/nP3RWy.

### Motivation

It seems unnecessary to convert `int` to `ptr` just to get its offset. In most cases, they generates the same assembly, but sometimes it may miss some optimizations since the analysis of `GEP` is not as perfect as that of arithmetic operation. One example is

https://github.com/dtcxzyw/llvm-opt-benchmark/blob/e3c822bf41df3a88ca38eba884a52b0cc7e70bf2/bench/protobuf/optimized/generated_message_reflection.cc.ll#L39860-L39873

``` llvm
  %conv.i188 = zext i32 %145 to i64
  %add.i189 = add i64 %conv.i188, %125
  %146 = load i16, ptr %num_aux_entries10.i, align 2
  %conv2.i191 = zext i16 %146 to i64
  %mul.i192 = shl nuw nsw i64 %conv2.i191, 3
  %add3.i193 = add i64 %add.i189, %mul.i192
  %147 = inttoptr i64 %add3.i193 to ptr
  %sub.ptr.lhs.cast.i195 = ptrtoint ptr %144 to i64
  %sub.ptr.rhs.cast.i196 = ptrtoint ptr %143 to i64
  %sub.ptr.sub.i197 = sub i64 %sub.ptr.lhs.cast.i195, %sub.ptr.rhs.cast.i196
  %add.ptr = getelementptr inbounds i8, ptr %147, i64 %sub.ptr.sub.i197
  %sub.ptr.lhs.cast = ptrtoint ptr %add.ptr to i64
  %sub.ptr.sub = sub i64 %sub.ptr.lhs.cast, %125
```

where `%conv.i188` first adds `%125` and then subtracts `%125` (the result is `%sub.ptr.sub`), which can be optimized.

But this fold involves `ptr`, which Alive2 can't verify, plus that in most cases, they generate the same assembly, I am not sure whether it is profitable.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp [hunk 1]
```diff
@@ -2073,6 +2073,16 @@ Instruction *InstCombinerImpl::visitPtrToInt(PtrToIntInst &CI) {
     return InsertElementInst::Create(Vec, NewCast, Index);
   }
 
+  // (ptrtoint (gep (inttoptr Base), Offset)) -> Base + Offset
+  Value *Base, *Offset;
+  if (match(SrcOp, m_OneUse(m_PtrAdd(m_OneUse(m_IntToPtr(m_Value(Base))),
+                                     m_Value(Offset)))) &&
+      Base->getType() == Ty && Offset->getType() == Ty) {
+    auto *NewOp = BinaryOperator::CreateAdd(Base, Offset);
+    NewOp->setHasNoUnsignedWrap(true);
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:2082]: This nuw is incorrect. You need inbounds gep + non-negative offset for the add to be nuw.

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp [hunk 2]
```diff
@@ -2073,6 +2073,20 @@ Instruction *InstCombinerImpl::visitPtrToInt(PtrToIntInst &CI) {
     return InsertElementInst::Create(Vec, NewCast, Index);
   }
 
+  // (ptrtoint (gep (inttoptr Base), Offset)) -> Base + Offset
+  Value *Base, *Offset, *GEP;
+  if (match(SrcOp, m_OneUse(m_Value(GEP))) &&
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:2078]: Add the `m_OneUse()` to the match below instead.

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp [hunk 3]
```diff
@@ -2073,6 +2073,20 @@ Instruction *InstCombinerImpl::visitPtrToInt(PtrToIntInst &CI) {
...
+      match(GEP,
+            m_PtrAdd(m_OneUse(m_IntToPtr(m_Value(Base))), m_Value(Offset))) &&
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:2080]: We could support this for arbitrary GEPs (not just ptradd) by calling `emitGEPOffset()` instead.

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp [hunk 4]
```diff
@@ -2073,6 +2073,20 @@ Instruction *InstCombinerImpl::visitPtrToInt(PtrToIntInst &CI) {
...
+      Base->getType() == Ty && Offset->getType() == Ty) {
+    auto *GEPInst = cast<GetElementPtrInst>(GEP);
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:2082]:
```suggestion
    auto *GEPInst = cast<GEPOperator>(GEP);
```
is safer

**Review by nikic — COMMENTED** [review]
This looks like a reasonable transform to me.

More generally, we could fold `ptrtoint(gep p, x)` to `add ptrtoint(p), x`, but I think this may currently interfere with some other things, so I'm fine with checking this specific pattern.

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp (hunk 1 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp]: @YanWQ-monad Can you provide the alive proof?

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp [hunk 5]
```diff
@@ -2049,16 +2049,28 @@ Instruction *InstCombinerImpl::visitPtrToInt(PtrToIntInst &CI) {
...
+      return replaceInstUsesWith(CI,
+                                 Builder.CreateIntCast(EmitGEPOffset(GEP), Ty,
+                                                       /*isSigned=*/false));
+    }
+
+    // (ptrtoint (gep (inttoptr Base), ...)) -> Base + Offset
+    Value *Base;
+    if (GEP->hasOneUse() &&
+        match(GEP->getPointerOperand(), m_OneUse(m_IntToPtr(m_Value(Base)))) &&
+        Base->getType() == Ty) {
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:2068]: Please add a negative test with mismatched types.

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp (hunk 1 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp]: You can treat `inbounds` as `nsw + nuw`. See also https://discourse.llvm.org/t/rfc-add-nusw-and-nuw-flags-for-getelementptr/78672.

Comment (YanWQ-monad) [reply llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp]: It's now added in the description. Thanks.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Comment by goldsteinn** [comment]
LGTM as well.

**Pull request merged**